### ISSUE TITLE
Feature Request: Generate Swagger Request Bodies from Validator Classes

### ISSUE THREAD

**Issue by HasanAshab** [ad-on-is/adonis-autoswagger#35]
Hello,

I am currently working on a project using AdonisJS and I am using the `adonis-autoswagger` package to generate Swagger documentation for my API. I have noticed that while `adonis-autoswagger` provides a lot of flexibility and control over the generated documentation, it does not currently support generating Swagger request bodies directly from validator classes.

This limitation has led to some redundancy in our codebase, as we have to manually define the structure of the request body in both the validator class and the Swagger documentation. This not only increases the maintenance effort but also makes the code harder to understand and manage. This will keep our docs sync the code base

I believe that adding support for generating Swagger request bodies directly from validator classes would greatly enhance the usability and efficiency of `adonis-autoswagger`. It would allow us to define the structure of the request body in one place (the validator class), and then automatically generate the corresponding Swagger documentation.

Here's a rough idea of how this could work:

```javascript


class LoginValidator {
  // generate the body using this property
  public schema = schema.create({
     email: schema.string(),
     password: schema.string()
  });
}

/**
 * @api {post} /login login a new user
 * @requestBody <LoginValidator>  // it will be equal to { email: string, password: string }
 */
Route.post('/login', async ({ request }) => {
 const user = await request.validate(LoginValidator)
 // ... rest of my code
})
```

In this example, the `LoginValidator` class defines the structure of the request body. We could then use this information to automatically generate the corresponding Swagger documentation for the `/login` endpoint.

I would appreciate if you could consider this feature request and let me know if there are any plans to implement it in the future. Alternatively, I would be happy to contribute to the development of this feature if you need help.

Thank you for considering this request.

Best regards,
Hasan Ashab

**Comment by ad-on-is** [comment]
Hi, thx for the suggestion. This would be a nice feature indeed.

For this, I'd need to parse the Validators/*.ts (like we already do with the models) so this should be easy enough. Unfortunately, I don't have the time, but PRs are more than welcome.

For now, you can use the **only()** or **without()** notations. Based on your schema, I suppose you have a model User, that has an email and passwor field. You can use `<User>.only(email, password)` or if the Validator has more fields, you can also use `<User>.without(avatar)`.

**Comment by HasanAshab** [comment]
Recently, i got work. so please anyone reading this issue, **contribute if you can**

Thanks in advance

**Comment by ad-on-is** [comment]
@HasanAshab You can however start working on this one if you'd like 😅, or do you think that the `only()` and `without()` notations are sufficient enough, so I can close it?

**Comment by HasanAshab** [comment]
Not sure, but I will start on it in some weeks Inshaallah :)

**Comment by ad-on-is** [comment]
May Allah bless you and give you the strength and patience 😅

**Comment by HasanAshab** [comment]
I can implement this by traversing validator properties but can't parse from text data 😅. It's pretty complex

**Comment by ad-on-is** [comment]
maybe the validatior has some method that returns a nicer object that can be easily parsed, in combination with dynamic imports using `eval()`.

thinking of smth like this

- read all validator files 
- get exported validatiors (regex "export const *Validator) => exampleValidator
- `const v = eval(import exampleValidator from currentfile.ts).then()`
- `v.toJson()` hopefully something exists
- OR `v.validate(null)` and hope that the object can be constructed from the error outputs

**Comment by HasanAshab** [comment]
Oh, I got the map

**Comment by HasanAshab** [comment]
@ad-on-is 
```typescript
const v = eval(import exampleValidator from currentfile.ts).then()
```
This line cause

```typescript
SyntaxError: Cannot use import statement outside a module
```

What about something like

```typescript
const validatorsDir = pkgJson.imports['#validators/*']
const paths = await AutoSwagger.default.getFiles(validatorsDir)
for(let path of paths) {
  path = path
    .replace('.ts', '')
    .replace('#validators', validatorsDir)

const validators = await import(path)

console.log(validators)
}
```

**Comment by HasanAshab** [comment]
@ad-on-is  I can make schema from vine schema. But found no way to make from compiled validators :/

**Comment by ad-on-is** [comment]
I'm not familiar with vinejs, but what do you get if you try to validate purposely wring data? do you get any useful structure?

```ts
validator.validate(null)
```

**Comment by HasanAshab** [comment]
Great! I can extract required field names from error obj
```typescript
[
    {                                                                   message: 'email field is required',                               rule: 'required',
      field: 'email'
    },
    {
      message: 'username field is required',
      rule: 'required',                                                 field: 'username'
    },
    {                                                                   message: 'password field is required',
      rule: 'required',
      field: 'password'                                               }
  ]
```

And also can extract their **types** by trying with all data types one by one.

But the problem is how to extract optional fields? have any Idea?

**Comment by ad-on-is** [comment]
> by trying with all data types one by one.

I think this might be too dirty and spaghetti. Is there a cleaner solution?

Does the validator provide any other useful methods or fields? does it provide the schema maybe?

**Comment by HasanAshab** [comment]
The **validator** don't save **schema** as prop when its compiled. Also found no cleaner solution.

Take a look on [VineValidator](https://github.com/vinejs/vine/blob/develop/src/vine/validator.ts#L36)

**Comment by HasanAshab** [comment]
But it may be quite difficult to parse raw text between **vine.compile()**. 

This may not work for some edge cases. Also dynamic validators would not be supported.

**Comment by ad-on-is** [comment]
You're right. I guess we might ditch this one after all, since it's technically impossible.

Or, as a final resort, I could ask the folks at adonisJS, whether they could provide a `toJSON()` function to the compiled validator, so we could get the map.

**Comment by HasanAshab** [comment]
Yeah please ask, I will wait for **toJSON()** to be added :)

**Comment by ad-on-is** [comment]
https://github.com/adonisjs/core/discussions/4480

Fingers crossed! 🤞🏻

**Comment by ad-on-is** [comment]
@HasanAshab 

Nice, now that we get a `toJSON()`, I'd suggest the following implementation.

```ts
// autoswagger.ts
// -----------------------------
 private async getSchemas() {
    let schemas = {
      Any: {
        description: "Any JSON object not defined as schema",
      },
    };

    schemas = {
      ...schemas,
      ...(await this.getInterfaces()),
      ...(await this.getModels()),
      ...(await this.getValidators()), // convert the validator.toJSON() to the same format as models/interfaces do
    };

    return schemas;
  }

private async getValidators() {

/* 
- read files
- use ValidatorParser for parsing
- return something like this 

{exampleValidator: {
type: "object",
properties: {}
description: "Validator"
},
otherValidator: {
type: "object",
properties: {}
description: "Validator"
}
}
*/

}


// parser.ts
// -----------------------------
export class ValidatorParser {
  // do your magic here
}

```

And the jsdoc notation should than work automatically with something like this
```
/* @requestBody <exampleValidator>

```

**Comment by HasanAshab** [comment]
I got it :) 👍

**Comment by ad-on-is** [comment]
Oh, one more thing.
Do you think it would be easy enough to get the errors of the validator as a response body?

Example (pseudo-code)
```ts
/**
   * @create
   * @requestBody <exampleValidator>
   * @responseBody 200 - <User>
   * @responseBody 422 - <exampleValidator.error>
   */

async create() {}
```

end then do something like this

```ts

let errors = []
try {
exampleValidator.validate({})
}catch(e) {
errors = e.message
}

```

**Comment by HasanAshab** [comment]
Thats good idea! I can generate multiple error response schemas from a validator, but does **AutoSwagger** supports multiple schema on same status code??

**Comment by ad-on-is** [comment]
No it does not support it, but I think it's sufficient enough just to have the schema of the error response as it is always the same.

**Comment by HasanAshab** [comment]
Okay and how can we know the exact type of **literal**?

**Comment by ad-on-is** [comment]
```ts
async index() {
    try {
      await testValidator.validate({
        name: 'John Doe',
        email: '',
      })
    } catch (e) {
      console.log(e.messages)
    }
    return {
      hello: 'world',
    }
  }
```

This is what I get as an output of console.log(e.messages)

```
[
  {
    message: 'The boing field must be defined',
    rule: 'required',
    field: 'boing'
  },
  {
    message: 'The title field must be defined',
    rule: 'required',
    field: 'title'
  },
  {
    message: 'The slug field must be defined',
    rule: 'required',
    field: 'slug'
  },
  {
    message: 'The description field must be defined',
    rule: 'required',
    field: 'description'
  }
]
```
Done! Nothing more is needed, just so everyone reading the docs knows, "oh ok, if I provide an invalid request-body, I'll get this kind of format as an error". You know what I mean? No need to be more complex here.

**Comment by HasanAshab** [comment]
Yeah, I got it.

**Comment by HasanAshab** [comment]
> If someone uses ` @responseBody 400 - <exampleValidator.error>` but the status of the error is _422_. It's okey to overwrite the status provided by the developer, since this is AutoSwagger 😅

I think we should respect the status provided by developer as he may customized response on exception handler, am I right?

**Comment by ad-on-is** [comment]
> See [schema types](https://github.com/vinejs/compiler?tab=readme-ov-file#schema-types) there is **literal** type for **string**, **number**, etc. Now how to know the exact type?

oh, that's what you mean by *literal*

Hmm.. I'm not exactly sure, but refs['ref1://1'].validator should be a function and hopefully provide more info.

**Comment by ad-on-is** [comment]
> > If someone uses ` @responseBody 400 - <exampleValidator.error>` but the status of the error is _422_. It's okey to overwrite the status provided by the developer, since this is AutoSwagger 😅
> 
> I think we should respect the status provided by developer as he may customized response on exception handler, am I right?

Yes, you're right. The devs should be responsible for setting a proper status code, in case they override it in the controller.

**Comment by HasanAshab** [comment]
How can I play around vinejs latest commit

**Comment by ad-on-is** [comment]
> How can I play around vinejs latest commit

I just did `pnpm i https://github.com/vinejs/vine#develop` and it seems to work. No need to use a specific commit, just the develop branch is sufficient

**Comment by HasanAshab** [comment]
Oh, thanks :)

**Comment by vithalreddy** [comment]
any progress on this?

Happy to help with testing this on big code base.

**Comment by ad-on-is** [comment]
Hey @HasanAshab, how is it going with this one?

**Comment by HasanAshab** [comment]
@ad-on-is @vithalreddy Sorry guys. I just migrated to Django :/. 

If anybody can, please **contribute** 🙏

**Comment by ad-on-is** [comment]
@HasanAshab ... all good, no pressure 👍

would you mind pushing a PR, with the state you currently have, so I or someone else can pick up where you left off

**Comment by HasanAshab** [comment]
I have no statable state. I just played with **.toJSON()** and some planning written on paper 😅

**Comment by ad-on-is** [comment]
Oh ok... I thought you were much further, with the dynamic imports, etc...

**Comment by RoyalBoogie** [comment]
Hi ! 
Unfortunately I can't help but I'll love to have this feature so I give you some strength ! 
Thanks btw for this package !

**Comment by antoninpire** [comment]
This feature would indeed be really cool, and I would not mind giving it a go.
However there seems to be an issue with the toJSON of the vine validators: all primary types are considered literal.
That means strings, numbers, booleans, enums etc are marked as "type": "literal" in the json returned.

**Comment by ad-on-is** [comment]
@antoninpire How did you get the latest vine working? or did they merge it into main?

**Comment by antoninpire** [comment]
I think they merged it into main, I didn't have to use a specific branch, the toJSON method was there

**Comment by ad-on-is** [comment]
@antoninpire Ooh cool, I will give it a shot then.

**Comment by ad-on-is** [comment]
Fixed in 3.45.0. I tried to do my best here. Feel free to test and report any issues or feedback.

Usage as always...

testValidator.ts
```ts

export const testValidator = vine.compile(
  vine.object({
    boing: vine.object({
      blub: vine.string(),
    }),
    title: vine.string().minLength(10).maxLength(80),
    mail: vine.string().email(),
    someArr: vine.array(vine.string()),
    somebool: vine.boolean(),
    complArray: vine.array(
      vine.object({
        best: vine.string(),
        test: vine.array(
          vine.object({
            final: vine.string(),
          })
        ),
      })
    ),
    enumChoice: vine.enum(['foo', 'bar']),
    description: vine.string().trim().escape(),
    data: vine.object({
      bar: vine.number().max(10).min(5),
      arr: vine.array(vine.number().min(10).max(20)),
      regex: vine.string().regex(/^[a-z]+$/),
      baz: vine.enum(['foo', 'bar']).nullable(),
      opts: vine.string().optional(),
    }),
  })
)
```

... and then...

```ts

@requestBody <testValidator>
```

**Comment by antoninpire** [comment]
Awesome! Thanks a lot
